core/drivers/ls_i2c.c

```diff
@@ -0,0 +1,404 @@
+
+	i2c_reset(ctrl_base);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_bus_test_bus_busy(struct i2c_regs *regs, bool test_busy)
+{
+	unsigned int n = 0;
+	uint8_t reg = 0;
+
+	for (n = 0; n < I2C_NUM_RETRIES; n++) {
+		reg = io_read8((vaddr_t)&regs->ibsr);
+
+		if (reg & I2C_IBSR_IBAL) {
+			io_write8((vaddr_t)&regs->ibsr, reg);
+			return TEE_ERROR_BUSY;
+		}
+
+		if (test_busy && (reg & I2C_IBSR_IBB))
+			break;
+
+		if (!test_busy && !(reg & I2C_IBSR_IBB))
+			break;
+
+		mdelay(1);
+	}
+
+	if (n == I2C_NUM_RETRIES)
+		return TEE_ERROR_BUSY;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_transfer_complete(struct i2c_regs *regs, bool test_rx_ack)
+{
+	unsigned int n = 0;
+	uint8_t reg = 0;
+
+	for (n = 0; n < I2C_NUM_RETRIES; n++) {
+		reg = io_read8((vaddr_t)&regs->ibsr);
+
+		if (reg & I2C_IBSR_IBIF) {
+			/* Write 1 to clear the IBIF field */
+			io_write8((vaddr_t)&regs->ibsr, reg);
+			break;
+		}
+		mdelay(1);
+	}
+
+	if (n == I2C_NUM_RETRIES)
+		return TEE_ERROR_BUSY;
+
+	if (test_rx_ack && (reg & I2C_IBSR_RXAK))
+		return TEE_ERROR_NO_DATA;
+
+	if (reg & I2C_IBSR_TCF)
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result i2c_read(struct i2c_regs *regs, unsigned int slave_address,
+			   struct i2c_operation *operation,
+			   bool is_last_operation)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int n = 0;
+
+	/* Write Slave Address */
+	io_write8((vaddr_t)&regs->ibdr, (slave_address << BIT(0)) | BIT(0));
+	res = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+	if (res)
+		return res;
+
+	/* select Receive mode. */
+	io_clrbits8((vaddr_t)&regs->ibcr, I2C_IBCR_TXRX);
+	if (operation->length_in_bytes > 1) {
+		/* Set No ACK = 0 */
+		io_clrbits8((vaddr_t)&regs->ibcr, I2C_IBCR_NOACK);
+	}
+
+	/* Perform a dummy read to initiate the receive operation. */
+	io_read8((vaddr_t)&regs->ibdr);
+
+	for (n = 0; n < operation->length_in_bytes; n++) {
+		res = i2c_transfer_complete(regs, I2C_BUS_NO_TEST_RX_ACK);
+		if (res)
+			return res;
+		if (n == (operation->length_in_bytes - 2)) {
+			/* Set No ACK = 1 */
+			io_setbits8((vaddr_t)&regs->ibcr, I2C_IBCR_NOACK);
+		} else if (n == (operation->length_in_bytes - 1)) {
+			if (!is_last_operation) {
+				/* select Transmit mode (for repeat start) */
+				io_setbits8((vaddr_t)&regs->ibcr,
+					    I2C_IBCR_TXRX);
+			} else {
+				/* Generate Stop Signal */
+				io_clrbits8((vaddr_t)&regs->ibcr,
+					    (I2C_IBCR_MSSL | I2C_IBCR_TXRX));
+				res = i2c_bus_test_bus_busy
+					(regs, I2C_BUS_TEST_IDLE);
+				if (res)
+					return res;
+			}
+		}
+		operation->buffer[n] = io_read8((vaddr_t)&regs->ibdr);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_write(struct i2c_regs *regs, unsigned int slave_address,
+			    struct i2c_operation *operation)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int n = 0;
+
+	/* Write Slave Address */
+	io_write8((vaddr_t)&regs->ibdr,
+		  (slave_address << BIT(0)) & (uint8_t)(~(BIT(0))));
```

Comment:
`BIT(0)` in `slave_address << BIT(0)` looks a bit funny, why are you using `BIT(0)` like this?
Instead of `(uint8_t)(~(BIT(0)))` I believe that `~BIT(0)` should be enough.

Reply:
Yes, Will update.